src/core/trie.js

```diff
@@ -0,0 +1,79 @@
+// @flow
+
+import {type AddressModule} from "./address";
+import {
+  NodeAddress,
+  type NodeAddressT,
+  EdgeAddress,
+  type EdgeAddressT,
+} from "./graph";
+import * as NullUtil from "../util/null";
+
+type Entry<T> = {map: RecursiveMap<T>, value: ?T};
+type RecursiveMap<T> = Map<string, Entry<T>>;
+class BaseTrie<K, T> {
+  addressModule: AddressModule<K>;
+  entry: Entry<T>;
+
+  constructor(m: AddressModule<K>) {
+    this.addressModule = m;
+    this.entry = {value: null, map: new Map()};
+  }
+
+  add(k: K, val: T): this {
+    const parts = this.addressModule.toParts(k);
+    let i = 0;
+    let entry = this.entry;
+    while (i < parts.length) {
```

Comment:
You can drop the `let i` and the `while`; this can be:

```js
for (const part of this.addressModule.toParts(k))
```

Reply:
Done